design/ceph/node-loss-rbd-rwo.md

```diff
@@ -0,0 +1,105 @@
+
+## Proposed Solution
+
+> Note: This solution requires minimum kubernetes version 1.26.0
+
+The kubernetes feature [Non-Graceful Node Shutdown](https://kubernetes.io/blog/2022/12/16/kubernetes-1-26-non-graceful-node-shutdown-beta/) is available starting in K8s 1.26 to help improve the volume recovery during node loss. When a node is lost, the admin is required to add the taint `out-of-service` manually to the node. After the node is tainted, K8s will:
+
+- Remove the volume attachment from the lost node
+- Delete the old pod on the lost node
+- Create a new pod on the new node
+- Allow the volume to be attached to the new node
+
+Once this taint is applied manually, rook will create a [NetworkFence CR](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md). The [csi-addons operator](https://github.com/csi-addons/kubernetes-csi-addons) will then blocklist the node to prevent any ceph rbd client on the lost node from writing any more data.
+
+After the new pod is running on the new node and the old node which was lost comes back, rook will delete the [NetworkFencr CR](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md).
+
+example of taint to be applied to lost node:
+
+```console
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoExecute
+# or
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoSchedule
+```
+
+Example of a NetworkFence CR that the Rook operator would create when a node is lost:
+
+```yaml
+apiVersion: csiaddons.openshift.io/v1alpha1
+kind: NetworkFence
+metadata:
+  name: <name> # We will keep the name the same as the node name
+  namespace: <ceph-cluster-namespace>
+spec:
+  driver: <driver-name> # for us it will `rook-ceph.rbd.csi.ceph.com`
+  fenceState: <fence-state> # For us it will be `Fenced`
+  cidrs:
+    - <nodeIP>
+  secret:
+    name: <csi-rbd-provisioner-secret-name>
+    namespace: <ceph-cluster-namespace>
+  parameters:
+    clusterID: <ceph-cluster-namespace>
+```
+
+> Note: This will be enabled by default in rook, in the case for some reason user wants to disable this feature in rook can edit `rook-ceph-operator-config` configmap and update the `ROOK_WATCH_FOR_NODE_FAILURE: "false"`.
+
+## How get which NodeIp to block
+
+There are multiple networking options available for example, host Networking, Pod networking, Multus etc. This make it difficult to know which NodeIp address to blockList.
+For this we'll follow the following approach which will work for all networking options(but still will not work for external cluster which nodeIps):
+
+1. Get the `volumesInUse` from the node which has the taint `out-of-service`
+
+2. List all the pv and compare with value of `volumesInUse`
+
+3. Once, above mataches:
+
+   1. If RBD PVC makes use of the rbd status API
```

Comment:
We will query `rbd status`? Or what do you mean we make use of the rbd status API or the `ceph fs client ls`? Please explain more how these are queried. Maybe it would clarify if we separated RBD and CephFS into separate sections.

Reply:
Maybe this https://github.com/rook/rook/issues/11372#issuecomment-1336809035 will explain all. Let me know if you want to add something specific

Reply:
I see you have an example on line 77 below. How about moving that example up here by this bullet, and then have a cephfs example by bullet 2?